### PR TITLE
Implement user permission tests

### PR DESCRIPTION
**Pull request by Accelsnow**
Working on `/frontend/tests/user-permissioin-tests.js`.

TODO-tests to be implemented:

1. An instructor can access instructor routes
2. A TA can access TA routes
3. An instructor who is also a TA can access instructor and TA routes

### CODE REVIEW

**Comment by Accelsnow** [comment]
`instructor/instructors` route gets `204 No Content` response from the ruby backend.

```
Started GET "/api/v1/instructor/instructors" for 172.21.0.6 at 2021-01-26 20:03:20 +0000
  User Load (0.1ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 316], ["LIMIT", 1]]
  ↳ app/services/active_user_service.rb:36:in `active_user'
Processing by Api::V1::Instructor::InstructorsController#index as HTML
  CACHE User Load (0.0ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 316], ["LIMIT", 1]]
  ↳ app/services/active_user_service.rb:36:in `active_user'
  Instructor Load (0.1ms)  SELECT "instructors".* FROM "instructors" WHERE "instructors"."utorid" = $1 LIMIT $2  [["utorid", "smithh"], ["LIMIT", 1]]
  ↳ app/controllers/api/v1/instructor/instructors_controller.rb:11:in `index'
Completed 204 No Content in 1ms (ActiveRecord: 0.2ms | Allocations: 1083)
```

This causes the frontend test prop type check to fail, since the type check expects an object as the HTTP response body, while 204 response body is an empty string.

```
● Console

    console.error node_modules/prop-types/checkPropTypes.js:20
      Warning: Failed PropType verification type: Invalid PropType verification `<root>` of type `string` supplied to `the client`, expected `object`.
    console.warn src/tests/utils.js:284
      Proptypes error call stack: Error: 
          at checkPropTypes (/app/src/tests/utils.js:282:19)
          at apiGET (/app/src/tests/utils.js:94:5)
          at runMicrotasks (<anonymous>)
          at processTicksAndRejections (internal/process/task_queues.js:97:5)
          at Object.<anonymous> (/app/src/tests/user-permission-tests.js:72:20)
```

File: frontend/src/tests/user-permission-tests.js [hunk 1]
```diff
@@ -14,11 +14,41 @@ import { databaseSeeder } from "./setup";
...
+        utorid: "tauser",
+        roles: ["ta"],
+    };
+
+    const newTAInstructorData = {
+        utorid: "tainstuser",
+        roles: ["ta", "instructor"],
+    };
+
+    async function runTestWithUser(userData, testBody) {
```
Comment (Accelsnow) [inline frontend/src/tests/user-permission-tests.js]: Need to parse this function into a setup-teardown style instead of passing a function into a function

File: frontend/src/tests/user-permission-tests.js [hunk 2]
```diff
@@ -14,11 +14,43 @@ import { databaseSeeder } from "./setup";
...
+        roles: ["ta"],
+    };
+
+    const newTAInstructorData = {
+        utorid: "tainstuser",
+        roles: ["ta", "instructor"],
+    };
+
+    async function setupUserPerm(userData) {
+        let respPrevUser = await apiGET("/active_user");
```
Comment (Altair59) [inline frontend/src/tests/user-permission-tests.js]: Maybe I am wrong, but I believe since respPrevUser is not changed we can use const here, as what is prevUserId

File: frontend/src/tests/user-permission-tests.js [hunk 3]
```diff
@@ -14,11 +14,43 @@ import { databaseSeeder } from "./setup";
 export function userPermissionsTests(api) {
     const { apiGET, apiPOST } = api;
 
-    const newUserData = {
+    const newInstructorData = {
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: This is user data, not instructor data...

File: frontend/src/tests/user-permission-tests.js [hunk 4]
```diff
@@ -14,11 +14,43 @@ import { databaseSeeder } from "./setup";
...
         utorid: "userxx",
         roles: ["instructor"],
     };
 
+    const newTAData = {
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: How about, `userTA` and `userInstructor`, etc. for variable names?

File: frontend/src/tests/user-permission-tests.js [hunk 5]
```diff
@@ -14,11 +14,43 @@ import { databaseSeeder } from "./setup";
...
+        utorid: "tauser",
+        roles: ["ta"],
+    };
+
+    const newTAInstructorData = {
+        utorid: "tainstuser",
+        roles: ["ta", "instructor"],
+    };
+
+    async function setupUserPerm(userData) {
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: Don't use abbreviations unless they're more common than the full word. E.g., `Perm`->`Permissions`

File: frontend/src/tests/user-permission-tests.js [hunk 6]
```diff
@@ -14,11 +14,43 @@ import { databaseSeeder } from "./setup";
...
+        const respPrevUser = await apiGET("/active_user");
+        expect(respPrevUser).toHaveStatus("success");
+        const prevUserId = respPrevUser.payload.id;
+
+        let resp = await apiPOST("/admin/users", userData);
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: A mix of `let` and `const`? Make a choice and stick with it.

File: frontend/src/tests/user-permission-tests.js (hunk 5 above)
Comment (siefkenj) [reply frontend/src/tests/user-permission-tests.js]: The name of this function doesn't clue me in on what's happening. Try again.

Comment (siefkenj) [reply frontend/src/tests/user-permission-tests.js]: Function is missing a doc string. All utility functions should have doc strings. (Look in `utils.js` for examples). You can install the `Document This` extension for VSCode to make creating doc strings easier.

File: frontend/src/tests/user-permission-tests.js [hunk 7]
```diff
@@ -30,26 +62,54 @@ export function userPermissionsTests(api) {
     });
 
     it("A non-admin cannot access the admin route", async () => {
-        let resp = await apiPOST("/admin/users", newUserData);
-        // Save any extra attributes we got, such as an `id`
-        Object.assign(newUserData, resp.payload);
-
-        // Set the active user to the newly-created user
-        resp = await apiPOST("/debug/active_user", { id: newUserData.id });
+        const prevUserId = await setupUserPerm(newInstructorData);
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: Based on this pattern, it might be nicer to have a class that is initialized with a user e.g. named `ActiveUserSwitcher`, and then you can call `await ActiveUserSwitcher.makeActive()` and `await ActiveUserSwitcher.revert()` or something like that.

Comment (siefkenj) [reply frontend/src/tests/user-permission-tests.js]: If you create a class for this, it should also have its own test (maybe an existing `active_user` can do double duty?)

File: frontend/src/tests/user-permission-tests.js [hunk 8]
```diff
@@ -30,26 +62,54 @@ export function userPermissionsTests(api) {
...
 
         // Try to fetch an admin route
-        resp = await apiGET("/admin/active_user");
+        let resp = await apiGET("/admin/active_user");
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js:85]: You test a single route...let's test many.

File: frontend/src/tests/user-permission-tests.js [hunk 9]
```diff
@@ -30,26 +62,54 @@ export function userPermissionsTests(api) {
...
-        resp = await apiPOST("/debug/active_user", {
-            id: databaseSeeder.seededData.active_user.id,
-        });
+        await restoreOriginalUserPerm(prevUserId);
+    });
+
+    it("An instructor can access instructor routes", async () => {
+        const prevUserId = await setupUserPerm(newInstructorData);
+
+        // the default user has all roles, including the instructor role
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js:110]: Pick a user with only the instructor role.

Comment (siefkenj) [reply frontend/src/tests/user-permission-tests.js:110]: Also test for a user with multiple roles.

File: frontend/src/tests/user-permission-tests.js [hunk 10]
```diff
@@ -30,26 +62,54 @@ export function userPermissionsTests(api) {
...
-        "An instructor who is also a TA can access instructor and TA routes"
-    );
+    it("A TA can access TA routes", async () => {
+        // the default user has all roles, including the TA role
+        let resp = await apiGET("/ta/active_user");
+        expect(resp).toHaveStatus("success");
+
+        const prevUserId = await setupUserPerm(newTAData);
+
+        // Try to fetch an instructor route
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: instructor -> ta

File: frontend/src/tests/user-permission-tests.js [hunk 11]
```diff
@@ -30,26 +62,54 @@ export function userPermissionsTests(api) {
...
+        expect(resp).toHaveStatus("success");
+        resp = await apiGET("/instructor/instructors");
+        expect(resp).toHaveStatus("success");
+
+        const prevUserId = await setupUserPerm(newTAInstructorData);
+
+        // Try to fetch an instructor route
+        resp = await apiGET("/ta/active_user");
+        expect(resp).toHaveStatus("success");
+        resp = await apiGET("/instructor/instructors");
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: Why are you using `/instructor/instructors` instead of `/instructor/active_user`?

Comment (Accelsnow) [reply frontend/src/tests/user-permission-tests.js:110]: can you explain the question a bit more? I used `/instructor/instructors` here just because it is an instructor route.

Comment (Accelsnow) [reply frontend/src/tests/user-permission-tests.js]: > Why are you using `/instructor/instructors` instead of `/instructor/active_user`?

can you explain the question a bit more? I used /instructor/instructors here just because it is an instructor route.

File: frontend/src/tests/user-permission-tests.js (hunk 10 above)
Comment (Accelsnow) [reply frontend/src/tests/user-permission-tests.js]: > instructor -> ta

i do not understand this review. would you please explain this more?

File: frontend/src/tests/user-permission-tests.js (hunk 11 above)
Comment (siefkenj) [reply frontend/src/tests/user-permission-tests.js]: `/instructor/active_user` is also an instructor route. You used `/ta/active_user`, so `/instructor/active_user` seems like the natural choice. You should have a reason if you deviate from the natural choice.

File: frontend/src/tests/user-permission-tests.js [hunk 12]
```diff
@@ -14,42 +14,127 @@ import { databaseSeeder } from "./setup";
...
+    };
+
+    /**
+     * Switch current active user to the user represented by the given user data. If given user does not exist, one will
+     * be created.
+     * @param userData the user to be set to the active user
+     * @returns {Promise<void>}
+     */
+    async function switchToTargetUser(userData) {
+        let resp = await apiPOST("/debug/users", userData);
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: You create a new user each time before you switch to it? That doesn't seem right...

File: frontend/src/tests/user-permission-tests.js [hunk 13]
```diff
@@ -14,42 +14,127 @@ import { databaseSeeder } from "./setup";
...
+        let respSwitchBackUser = await apiPOST(
+            `/debug/active_user`,
+            defaultUser
+        );
+        expect(respSwitchBackUser).toHaveStatus("success");
+    }
+
     beforeAll(async () => {
         await databaseSeeder.seed(api);
+        const respOriginalUser = await apiGET(`/debug/active_user`);
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: Just use `resp`

File: frontend/src/tests/user-permission-tests.js [hunk 14]
```diff
@@ -14,42 +14,127 @@ import { databaseSeeder } from "./setup";
...
     }, 30000);
 
+    afterEach(async () => {
+        await restoreDefaultUser();
+    }, 3000);
+
     it("Admin user can access admin route", async () => {
         // the default user has all roles, including the admin role
-        const resp = await apiGET("/admin/active_user");
+        let resp = await apiGET("/admin/active_user");
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js:77]: Why did you change this?

File: frontend/src/tests/user-permission-tests.js [hunk 15]
```diff
@@ -14,42 +14,127 @@ import { databaseSeeder } from "./setup";
 export function userPermissionsTests(api) {
     const { apiGET, apiPOST } = api;
 
-    const newUserData = {
+    let defaultUser;
+
+    const newInstructorOnlyUserData = {
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: Pick a better name. Why is the word `new` in it?

File: frontend/src/tests/user-permission-tests.js (hunk 10 above)
Comment (siefkenj) [reply frontend/src/tests/user-permission-tests.js]: The comment says the word "instructor" but you mean to have the word "ta"

File: frontend/src/tests/user-permission-tests.js [hunk 16]
```diff
@@ -14,42 +14,127 @@ import { databaseSeeder } from "./setup";
...
+        const respOriginalUser = await apiGET(`/debug/active_user`);
+        expect(respOriginalUser).toHaveStatus("success");
+        defaultUser = {
+            utorid: respOriginalUser.payload.utorid,
+            roles: respOriginalUser.payload.roles,
+        };
     }, 30000);
 
+    afterEach(async () => {
+        await restoreDefaultUser();
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: It seems weird to have an `afterEach` with no `beforeEach`. It makes it hard to follow exactly what's going on in each test. Just repeat the one line each time you need it.

**Comment by siefkenj** [comment]
Failing code formatting tests....

File: frontend/src/tests/user-permission-tests.js [hunk 17]
```diff
@@ -14,42 +14,142 @@ import { databaseSeeder } from "./setup";
...
-    const newUserData = {
-        utorid: "userxx",
+    let defaultUser;
+
+    const instructorOnlyUserData = {
+        utorid: "user_permission_test_instructor_only_utorid",
         roles: ["instructor"],
     };
 
+    const TAOnlyUserData = {
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: lower case `ta`. There's no need for the word `Data` to be in this variable name.

File: frontend/src/tests/user-permission-tests.js [hunk 18]
```diff
@@ -14,42 +14,142 @@ import { databaseSeeder } from "./setup";
...
+        roles: ["ta", "instructor"],
+    };
+
+    /**
+     * Switch current active user to the user represented by the given user data. If given user does not exist, one will
+     * be created.
+     * @param newActiveUser the user to be set to the active user
+     * @returns {Promise<void>}
+     */
+    async function switchToTargetUser(newActiveUser) {
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: No need for word `Target`

File: frontend/src/tests/user-permission-tests.js [hunk 19]
```diff
@@ -14,42 +14,142 @@ import { databaseSeeder } from "./setup";
...
+        expect(resp).toHaveStatus("success");
+    }
+
+    /**
+     * Restores the active user to the default user (the user logged during test setup in beforeAll).
+     *
+     * @returns {Promise<void>}
+     */
+    async function restoreDefaultUser() {
+        let respSwitchBackUser = await apiPOST(
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: If you're not doing something meaningful with the response objects, just name them `resp`.

File: frontend/src/tests/user-permission-tests.js [hunk 20]
```diff
@@ -14,42 +14,142 @@ import { databaseSeeder } from "./setup";
...
     beforeAll(async () => {
         await databaseSeeder.seed(api);
+        let resp = await apiGET(`/debug/active_user`);
+        expect(resp).toHaveStatus("success");
+        defaultUser = {
+            utorid: resp.payload.utorid,
+            roles: resp.payload.roles,
+        };
+
+        resp = await apiPOST("/debug/users", instructorOnlyUserData);
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: Comment saying why all these calls are being made.

File: frontend/src/tests/user-permission-tests.js [hunk 21]
```diff
@@ -14,42 +14,142 @@ import { databaseSeeder } from "./setup";
...
+        expect(resp).toHaveStatus("error");
+
+        resp = await apiGET("/admin/applicants");
+        expect(resp).toHaveStatus("error");
+
+        await restoreDefaultUser();
+    });
+
+    it("An instructor can access instructor routes", async () => {
+        // the default user has all roles, including the instructor role
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: Are we sure we're currently the default user? I would feel safer with a `restoreDefaultUser()` before this.

File: frontend/src/tests/user-permission-tests.js [hunk 22]
```diff
@@ -14,42 +14,142 @@ import { databaseSeeder } from "./setup";
...
-    it.todo("An instructor can access instructor routes");
-    it.todo("A TA can access TA routes");
-    it.todo(
-        "An instructor who is also a TA can access instructor and TA routes"
-    );
+    it("An instructor who is also a TA can access instructor and TA routes", async () => {
+        // the default user has all roles, including both the instructor role and the TA role
+        let resp = await apiGET("/ta/active_user");
+        expect(resp).toHaveStatus("success");
+        resp = await apiGET("/instructor/instructors");
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: Again, `/instructor/instructors` is a weird route to be checking when you have `/instructor/active_user` that you could check instead.

File: frontend/src/tests/user-permission-tests.js [hunk 23]
```diff
@@ -4,52 +4,148 @@ import { databaseSeeder } from "./setup";
...
     beforeAll(async () => {
         await databaseSeeder.seed(api);
+        let resp = await apiGET(`/debug/active_user`);
+        expect(resp).toHaveStatus("success");
+        defaultUser = {
+            utorid: resp.payload.utorid,
+            roles: resp.payload.roles,
+        };
+
+        // create in the database the user with only instructor role
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: It's clear what the code is doing, so your comment doesn't need to address that. What you comment should address is *why* you're doing it.

File: frontend/src/tests/user-permission-tests.js [hunk 24]
```diff
@@ -4,52 +4,148 @@ import { databaseSeeder } from "./setup";
...
-        resp = await apiPOST("/debug/active_user", {
-            id: databaseSeeder.seededData.active_user.id,
-        });
+        await restoreDefaultUser();
+    });
+
+    it("An instructor can access instructor routes", async () => {
+        // switch to instructor only user
+        await switchToUser(instructorOnlyUser);
+        let resp = await apiGET("/instructor/instructors");
```
Comment (siefkenj) [inline frontend/src/tests/user-permission-tests.js]: Switch these to `/instructor/active_user`

**Comment by siefkenj** [comment]
Looks great!
